`from_py_with` doesn't seem to play nicely with optional arguments

### Bug Description

When trying to build optional arguments, using `from_py_with` I'm getting 

```
expected struct `InnerType`, found enum `Option`
```

### Steps to Reproduce

```rs

fn build_regex(obj: &PyAny) -> PyResult<Option<Regex>> {
    if obj.is_none() {
        Ok(None)
    } else {
        let regex_str: &str = obj.extract()?;
        let regex = Regex::new(regex_str).map_err(|e| PyValueError::new_err(e.to_string()))?;
        Ok(Some(regex))
    }
}

#[pyfunction]
pub fn regex_match(
    str: String,
    #[pyo3(from_py_with = "build_regex")] pattern: Option<Regex>,
) -> PyResult<bool> {
    if let Some(pattern) = pattern {
        if !pattern.is_match(&str) {
            return Ok(false);
        }
    }
    Ok(true)
}
```

(I know in this case I could build the regex inside `regex_match` but in practice I want to be able to call `regex_match` from other code with a pre-compiled regex.

### Backtrace

```shell
error[E0308]: mismatched types
  --> src/validators.rs:36:52
   |
36 |     #[pyo3(from_py_with = "build_regex")] pattern: Option<Regex>,
   |                                                    ^^^^^^ expected struct `regex::Regex`, found enum `Option`
   |
   = note: expected struct `regex::Regex`
                found enum `Option<_>`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `pydantic-core` due to previous error
```
```


### Your operating system and version

MacOS 12.2.1

### Your Python version (`python --version`)

3.10.4

### Your Rust version (`rustc --version`)

rustc 1.59.0 (9d1b2106e 2022-02-23)

### Your PyO3 version

`{version = "0.16.3", features = ["extension-module"]}`

### How did you install python? Did you use a virtualenv?

pyenv, yes

### Additional Info

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - Fix visibility of `PyDictItems`, `PyDictKeys`, and `PyDictValues` types added in PyO3 0.17.0.
+- Fix compile failure when using `#[pyo3(from_py_with = "...")]` attribute on an argument of type `Option<T>`. [#2592](https://github.com/PyO3/pyo3/pull/2592)
 
 ## [0.17.0] - 2022-08-23
 

--- a/pyo3-macros-backend/src/params.rs
+++ b/pyo3-macros-backend/src/params.rs
@@ -247,7 +247,7 @@ fn impl_arg_param(
             }
             (None, true) => {
                 quote_arg_span! {
-                    _pyo3::impl_::extract_argument::from_py_with_with_default(#arg_value, #name_str, #expr_path, || Some(None))?
+                    _pyo3::impl_::extract_argument::from_py_with_with_default(#arg_value, #name_str, #expr_path, || None)?
                 }
             }
             (None, false) => {

--- a/tests/test_pyfunction.rs
+++ b/tests/test_pyfunction.rs
@@ -165,6 +165,42 @@ fn test_function_with_custom_conversion_error() {
     });
 }
 
+#[test]
+fn test_from_py_with_defaults() {
+    fn optional_int(x: &PyAny) -> PyResult<Option<i32>> {
+        if x.is_none() {
+            Ok(None)
+        } else {
+            Some(x.extract()).transpose()
+        }
+    }
+
+    // issue 2280 combination of from_py_with and Option<T> did not compile
+    #[pyfunction]
+    fn from_py_with_option(#[pyo3(from_py_with = "optional_int")] int: Option<i32>) -> i32 {
+        int.unwrap_or(0)
+    }
+
+    #[pyfunction(len = "0")]
+    fn from_py_with_default(#[pyo3(from_py_with = "PyAny::len")] len: usize) -> usize {
+        len
+    }
+
+    Python::with_gil(|py| {
+        let f = wrap_pyfunction!(from_py_with_option)(py).unwrap();
+
+        assert_eq!(f.call0().unwrap().extract::<i32>().unwrap(), 0);
+        assert_eq!(f.call1((123,)).unwrap().extract::<i32>().unwrap(), 123);
+        assert_eq!(f.call1((999,)).unwrap().extract::<i32>().unwrap(), 999);
+
+        let f2 = wrap_pyfunction!(from_py_with_default)(py).unwrap();
+
+        assert_eq!(f2.call0().unwrap().extract::<usize>().unwrap(), 0);
+        assert_eq!(f2.call1(("123",)).unwrap().extract::<usize>().unwrap(), 3);
+        assert_eq!(f2.call1(("1234",)).unwrap().extract::<usize>().unwrap(), 4);
+    });
+}
+
 #[pyclass]
 #[derive(Debug, FromPyObject)]
 struct ValueClass {

